lib/rpush/daemon/apnsp8/delivery.rb

```diff
@@ -8,6 +8,7 @@ module Apnsp8
       class Delivery < Rpush::Daemon::Delivery
         RETRYABLE_CODES = [ 429, 500, 503 ]
         CLIENT_JOIN_TIMEOUT = 60
+        DEFAULT_MAX_CONCURRENT_STREAMS = 100
```

Comment:
Not sure what a reasonable default is, but this seemed like a good place to start. I guess we could also add a new config for it? Or alternatively, use the batch size config? Thoughts?

Reply:
Maybe @mariannegru can shed some light on where and how the `0x7fffffff` was supposed to work. As per the comment in the code, it does come from the `http-2` gem:
https://github.com/igrigorik/http-2/blob/v0.11.0/lib/http/2/connection.rb#L15

But there's also:
https://github.com/igrigorik/http-2/blob/v0.11.0/lib/http/2/connection.rb#L24

So maybe 100 is indeed the "correct" value here.

Reply:
The max current streams gets read from the server, but if the server doesn't provide that header, http2 uses `0x7fffffff` as the default, which is supposed to behave like "infinity".

That default of 100 in http2 is for the local settings vs the server settings. So that's another possible approach is to look at local settings if server settings have the default of `0x7fffffff`

Reply:
> That default of 100 in http2 is for the local settings vs the server settings.

Ah, thanks for clarifying, I missed that.

Just to clarify, if the http2 gem was configured to use "infinite" connections, we internally mapped that to 0 and this resulted in the bug. So I assume 100 being infinitely less than infinity but bigger than 0, this would be okay, but I have no deep knowledge of the inner workings of this gem.

> I guess we could also add a new config for it?

As for making it configurable, I have no opinion on this, since while we use http2 at work, we don't use the http2 gem directly. I assume being able to pass your own http2 client would be nice, since this is currently a private method, if I understood correctly.

https://github.com/rpush/rpush/blob/v5.4.0/lib/rpush/daemon/dispatcher/apns_http2.rb#L33-L41

A good first step is probably getting rid of this bug and if people feel the need to configure it, discuss it then and how we'd expose it.

> So that's another possible approach is to look at local settings if server settings have the default of 0x7fffffff

I assume that would again be `100` (or more specifically, the default local value of the http2 gem), since we don't let users of Rpush configure this value. Since the http2 client isn't configurable at all, using the default values of the http2 gem actually sounds reasonable.

Did I understand this correctly?

Reply:
> Did I understand this correctly?

Yep, that all sounds right to me!

> using the default values of the http2 gem actually sounds reasonable

So just to make sure I'm understanding, you're voting for the approach to check local settings if the server settings are set to `0x7fffffff`? Or are you saying, leave it hard-coded at 100, which matches the default local settings?

Reply:
> So just to make sure I'm understanding, you're voting for the approach to check local settings if the server settings are set to `0x7fffffff`? Or are you saying, leave it hard-coded at 100, which matches the default local settings?

Good point. I meant it as a question, kind of. If we can, I'd say let's use the local settings. That way we don't have to decide this ourselves but rather trust the http2-gems defaults. WDYT?

Reply:
Yep, I'm on board with that 👍  I'll play around with it later this afternoon.

Reply:
Oh oops, I take it back, it's a bit more complicated than that...  The `NetHttp2::Client` doesn't expose `#local_settings`. It wraps an `HTTP2::Client`, which has both `#remote_settings` and `#local_settings`, but the `NetHttp2::Client` only exposes `#remote_settings` for some reason. To get to `#local_settings`, we'd need to get a change pushed through on [net-http2](https://github.com/ostinelli/net-http2).

So I guess now I'm back to voting for the simple, hard-coded `100` approach.

What does everyone think?

Reply:
If we can't access the local settings, hard-coding to 100 with a comment that basically repeats your comment ("ideally we'd use `local_settings` but... so we use 100.") is good enough for now, as it fixes this infinite loop.

Reply:
Comment added, thanks!